### ISSUE TITLE
Take nc_filter_actions() out of dispatch table, and add nc_def_var_filter() and nc_inq_var_filter() instead, keeping nc_filter_actions() internal to netcdf-c

### ISSUE THREAD

**Issue by edwardhartnett** [Unidata/netcdf-c#1836]
The filter_actions function in the dispatch table requires some constants from ncfilter.h, which is not installed.

So I will have to copy these constants to PIO code in order to have them, which is not ideal. Either ncfilter.h needs to be installed, or else these constants (and perhaps everything in ncfilter.h) needs to be moved to netcdf_filter.h, which is installed.

The constants are:
```
/* Internal filter actions */
#define NCFILTER_DEF		1
#define NCFILTER_REMOVE  	2
#define NCFILTER_FILTERIDS      3
#define NCFILTER_INFO		4
#define NCFILTER_CLIENT_REG	10
#define NCFILTER_CLIENT_UNREG	11
#define NCFILTER_CLIENT_INQ	12

```

**Comment by DennisHeimbigner** [comment]
I see Charlie Zender is also having a problem.
I need to rethink this. I have no problem putting those
constants in netcdf_filter, but there are some corresponding
data structures that I think are too complicated and it would
be undesirable to put them in netcdf_filter.

**Comment by WardF** [comment]
Thanks Ed & Dennis.  The coursework has lightened up finally so I'm catching up on PR's and other dev issues.  Dennis, do we need to move data structures into the private header? Or can we just move the constants?

**Comment by edhartnett** [comment]
If we are taking a look at this again, I will say that having a void * in
the dispatch table is a LOT of work for users of the dispatch table.

Basically, all the related PIO functions will have to unpack the void into
known types, based on filter ID. Filter IDs that PIO is unfamiliar with
will be impossible to handle, because it's impossible to know how many
bytes are in the void pointer struct!

So it would have been a lot simpler to have three or four filter functions,
each of which used only known parameter types. Failing that, if we do need
a void pointer, another argument telling how many bytes are in the
void pointer would help a lot.

Ed

**Comment by edwardhartnett** [comment]
OK, now I have encountered the problem that Charlie probably was mentioning, and it's a big one, which will not be best solved with making these structs public. It seems that loading the "args" parameter of nc_filter_actions() is challenging. Here's some code that does it, if I am reading it correctly (from libhdf5/hdf5filter.c):

```
        if(obj->usort != NC_FILTER_UNION_SPEC)
	    {stat = THROW(NC_EFILTER); goto done;}
        for(found=0,k=0;k<nfilters;k++) {
	    NC_FILTERX_SPEC* f = (NC_FILTERX_SPEC*)nclistget(var->filters,k);
	    if(strcmp(f->filterid,obj->u.spec.filterid)==0) {
	        obj->u.spec.nparams = f->nparams;
		if(f->params != NULL && f->nparams > 0) {
		    if((stat=NC_filterx_copy(f->nparams,(const char**)f->params,&obj->u.spec.params))) goto done;
		}
		found = 1;
		break;
	    }
	}
```

So we see this uses a function NC_filterx_copy() and some new structs (NC_FILTER_UNION_SPEC, NC_FILTERX_SPEC).

I think this is more than we want exposed, just to support a void argument in the dispatch table.

I suggest the following: the filter_actions function is the back end for just two other public filter functions: nc_inq_var_filter() and nc_def_var_filter(). Let's put those two functions in the dispatch table, instead of nc_filter_actions() (which can remain as an internal function of netcdf-c). In this way, the whole process of packing and unpacking the args parameter would be internal to netCDF, and no additional structs or pre-processor symbols need be exported in headers.

Basically I am struggling with how to support nc_filter_actions() in the dispatch table. Without all the extra defines and structs, it can't be used like other functions in the dispatch table. The dispatch table only seems to work well when each argument is a known C type. ;-(

**Comment by DennisHeimbigner** [comment]
I think you have a good point. As an approach to trying to keep the dispatch table
from exploding, it works, but the programming cost is probably too high.

**Comment by edwardhartnett** [comment]
OK, do you mind if I give this a try tomorrow? This is preventing any progress with getting PIO to handle new filter methods, which is of course required for new compression methods.

BTW there is a lot of interest in new compression methods in PIO, so this will be a welcome feature.

**Comment by DennisHeimbigner** [comment]
BTW, for clarity. You are using the user-defined format (UDF) for PIO, correct?
Are you using it for anything else?

**Comment by edwardhartnett** [comment]
Correct using UDF for PIO. I'm not using UDF anywhere else yet.

**Comment by DennisHeimbigner** [comment]
I have pushed the PR for this at https://github.com/Unidata/netcdf-c/pull/1852
